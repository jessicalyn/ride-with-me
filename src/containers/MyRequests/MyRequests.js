import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { Query, Mutation } from 'react-apollo';
import { gql } from 'apollo-boost';
import { Loader } from '../../components/Loader/Loader';
import { mutateRejectRequest } from '../../thunks/mutateRejectRequest';

const MY_REQUESTS = gql`
    query ($driverUuid: String!) {
      pendingRequests(driverUuid: $driverUuid) {
        id
        ride { id departureDate availableSeats startCity { name } endCity { name } }
        passenger { id firstName lastName }
        message
        status
        createdAt
      }
    }
  `

const ADD_RIDER = gql`
  mutation($passengerId: Int! $rideId: Int!) {
    createRidePassenger(passengerId:$passengerId, rideId:$rideId){
      ok
      message
    }
  }
`

const REJECT_RIDER = gql`
  mutation($id: Int!, $status: String!) {
    changeRequestStatus(id:$id status:$status){
      request {
        id
        status
      }
    }
  }
`

export class MyRequests extends Component {

  render() {
    const driverUuid = this.props.user.uuid
    return(
      <div className="containers">
        <h2>My Requests</h2>
        <Query
          query={MY_REQUESTS} variables={{ driverUuid }}
        >
        {({ loading, error, data }) => {
          if (loading) return <Loader />
          if (error) return <p>Error :(</p>
            
            return data.pendingRequests.map(({ id, ride, passenger, message, status, createdAt }) => (
                <div key={id} className="ride-container">
                  <h4>Request from { passenger.firstName }</h4>
                  <h4>To join your ride on { ride.departureDate } from { ride.startCity.name } to {ride.endCity.name }</h4>
                  <h4>Message: { message }</h4>
                  <h4>There are { ride.availableSeats } seats available.</h4>
                  <Mutation mutation={ADD_RIDER} variables={{ 
                    "passengerId": passenger.id,
                    "rideId": ride.id
                  }}>
                    {(createRidePassenger, { data, loading, error }) => {
                      if (loading) return <Loader />
                      if (error) return <p>{ error }</p> 
                      if (!data) return <button onClick={createRidePassenger}>Add { passenger.firstName } to Ride</button>
                      if (data) return <h3>Rider Added!</h3>
                    }
                  }
                  </Mutation>
                  <Mutation mutation={REJECT_RIDER} variables={{ 
                    "id": id,
                    "status": "rejected"
                  }}>
                    {(changeRequestStatus, { data, loading, error }) => {
                      if (loading) return <Loader />
                      if (error) return <p>{ error }</p>
                      if (!data) return <button onClick={changeRequestStatus}>Reject { passenger.firstName }'s Request</button>
                      if (data) return <h3>Rider Rejected!</h3>
                    }
                  }
                  </Mutation>
                </div>
                ))
              }}
        </Query>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  handleMutateReject: (url) => dispatch(mutateRejectRequest(url))
})

export default connect(mapStateToProps, mapDispatchToProps)(MyRequests)