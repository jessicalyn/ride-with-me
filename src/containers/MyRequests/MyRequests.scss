.requests-container {
  display: block;
  box-shadow: $grey-box-shadow;
}

.green-button {
  width: 50%
}

.red-button {
  width: 50%
}

.green-button:hover {
  background: #176F40;
  border-color: #176F40
}

.red-button:hover {
  background: #C21915;
  border-color: #C21915
}