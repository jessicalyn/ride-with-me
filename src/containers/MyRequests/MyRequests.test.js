import { MyRequests, mapStateToProps } from './MyRequests';
import React from 'react';
import { shallow } from 'enzyme';

describe('MyRequests Container', () => {
  describe('MyRequests', () => {
    let wrapper
    const mockUser = {id: 1, name: "Jessica"}

    beforeEach(() => {
      wrapper = shallow(
        <MyRequests 
          user={mockUser}
        />
      )
    })

    it('should match the snapshot', () => {
      expect(wrapper.debug()).toMatchSnapshot()
    })
  })

  describe('mapStateToProps', () => {
    it('should return the expected props to state', () => {

      const mockState = {
        user: {id: 1, name: "Jessica"},
        fakeState: "Not real state to return"
      }
      const expected = {
        user: {id: 1, name: "Jessica"}
      }
  
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })
  
})