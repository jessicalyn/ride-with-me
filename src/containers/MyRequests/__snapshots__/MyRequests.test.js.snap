// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyRequests Container MyRequests should match the snapshot 1`] = `
"<div className=\\"containers\\">
  <h2>
    My Requests
  </h2>
  <Query query={{...}} variables={{...}}>
    [function]
  </Query>
</div>"
`;
